libs/tx-builder/src/utils/deepDecoding.ts

```diff
@@ -0,0 +1,275 @@
+    proposalActions.map(async (action) => {
+      try {
+        return await decodeAction(options, action);
+      } catch (e) {
+        return {
+          error: true,
+          message: (e as Error).message,
+          data,
+        };
+      }
+    })
+  );
+
+  return decodedProposalActions;
+};
+
+type FunctionDataWithInputsReturnType = {
+  functionName: string;
+  inputs: {
+    name: string;
+    type: string;
+    value: string;
+  }[];
+};
+
+const decodeValue = (value: unknown): string => {
+  if (typeof value === 'string') {
+    if (value.startsWith('0x')) {
+      return fromHex(value as `0x${string}`, 'bigint').toString();
+    }
+    return value;
+  }
+  if (typeof value === 'number' || typeof value === 'bigint') {
+    return BigInt(value).toString();
+  }
+  return '0';
+};
+
+const decodeFunctionDataWithInputs = (options: {
+  abi: any[];
+  data: `0x${string}`;
+}): FunctionDataWithInputsReturnType => {
+  const result = decodeFunctionData(options);
+
+  const functionDetails = getAbiItem({
+    abi: options.abi,
+    name: result.functionName,
+  });
+
+  const inputs = functionDetails['inputs'] || [];
+
+  const inputsWithValues = (inputs as any[]).map((input, index) => ({
+    name: input.name,
+    type: input.type,
+    value: result.args?.[index]?.toString() || '0x',
+  }));
+
+  return {
+    functionName: result.functionName,
+    inputs: inputsWithValues,
+  };
+};
+
+const decodeAction = async (
+  options: Options,
+  action: MetaTransaction
+): Promise<DeepDecodedAction | ActionError> => {
+  const { data, to, value, operation } = action;
+  const { chainId, rpcs, loader } = options;
+  if (
+    !data ||
+    data === '0x' ||
+    !data.startsWith('0x') ||
+    data === ENCODED_0X0_DATA ||
+    (await getCode({ chainId, contractAddress: action.to, rpcs })) === '0x'
+  ) {
+    return {
+      to: to,
+      operation: operation || OperationType.Call,
+      name: `${HAUS_NETWORK_DATA[chainId]?.symbol} Transfer`,
+      value: decodeValue(value),
+      params: [],
+      decodedActions: [],
+    };
+  }
+
+  const { abi } = await whatsabi.autoload(to, {
+    provider: new providers.JsonRpcProvider(rpcs[chainId]),
+    followProxies: true,
+    abiLoader: loader,
+  });
+
+  if (!abi || !abi?.length) {
+    throw new Error('No ABI found for this contract');
+  }
+
+  const result = decodeFunctionDataWithInputs({
+    abi,
+    data: data as `0x${string}`,
+  });
+
+  if (!result) {
+    throw new Error('Could not decode action');
+  }
+
+  if (result.functionName === 'multiSend' && result.inputs.length === 1) {
+    const decodedActions = await decodeMultiCall(
+      options,
+      data as `0x${string}`
+    );
+
+    return {
+      to: to,
+      operation: OperationType.DelegateCall,
+      name: result.functionName,
+      value: decodeValue(value),
+      params: result.inputs,
+      decodedActions,
+    };
+  }
+
+  if (result.functionName.toLowerCase().includes('exec')) {
```

Comment:
I'm not very happy with how this is done, so please critique this and suggest better ways to solve this.

One clean way I can think of is to support only known functions such as `execTransactionFromModule` and slowly add support for similar functions in the wild in future patches.

Reply:
Yeah that is tough.

I like your idea of supporting the known functions. Maybe we could make a constant with the known function names and do the if statement on that? Would be easy then to add more.

```js
const knownFunctionsNeedsBetterName = ["execTransactionFromModule", "executeAsBaal"];

...

if (knownFunctionsNeedsBetterName.some(functionName => functionName === result.functionName)) {};
```

What do you think?

Reply:
@skuhlmann I've pushed a change to handle this more cleanly. Let me know what you think!